File: src/main/java/com/machfour/macros/cli/Help.java
package com.machfour.macros.cli;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import static com.machfour.macros.cli.CliMain.PROGNAME;

class Help extends CommandImpl {
    private static final String NAME = "help";
    private static final String USAGE = String.format("%s %s <command>", PROGNAME, NAME);

    Help() {
        super(NAME, USAGE);
    }

    @Override
    public void printHelp() {
        out.println("################################");
        out.println("## Max's nutrition calculator ##");
        out.println("################################");
        out.println();
        out.println("Available commands:");
        for (Command m : Commands.COMMANDS) {
            if (m.isUserCommand()) {
                out.println(m.name());
            }
        }
        out.println();
        out.printf("For help using a particular command, run %s %s <command> or %s <command> --help\n", PROGNAME, NAME, PROGNAME);
    }

    @Override
    public void doAction(List<String> args) {
        // help on a particular action
        if (args.size() >= 2 && Commands.CMDS_BY_NAME.containsKey(args.get(1))) {
            Command forHelp = Commands.CMDS_BY_NAME.get(args.get(1));
            forHelp.printHelp();
        } else {
            printHelp();
        }
    }
}

File: macros-cli/src/main/kotlin/com/machfour/macros/cli/modes/Help.kt
package com.machfour.macros.cli.modes

import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.commands
import com.machfour.macros.cli.commandsByName
import com.machfour.macros.core.CliConfig

class Help(config: CliConfig) : CommandImpl(config) {
    override val name = "help"
    override val usage = "Usage: ${config.programName} $name"

    override fun printHelp() {
        println("################################")
        println("## Max's nutrition calculator ##")
        println("################################")
        println()
        println("Available commands:")
        commands.forEach {
            if (it.isUserCommand) {
                println(it.name)
            }
        }
        println()
        println("For help using a particular command, run ${config.programName} $name <command>" +
                " or ${config.programName} <command> --$name")
    }

    override fun doAction(args: List<String>): Int {
        if (args.size >= 2) {
            // help on a particular action, if that action exists. Otherwise default help is printed
            commandsByName[args[1]]?.printHelp() ?: printHelp()
        } else {
            // no command specified
            printHelp()
        }
        return 0
    }

}